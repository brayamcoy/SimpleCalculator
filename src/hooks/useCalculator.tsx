import { ToastAndroid } from "react-native";
import React from "react";
import * as math from "mathjs";

export interface useCalculatorProps {
  value: string;
  setValue: React.Dispatch<React.SetStateAction<string>>;
  operation: string;
  setOperation: React.Dispatch<React.SetStateAction<string>>;
}

/**
 * useCalculator.tsx
 * Use Calculator hooks logic for the calculator
 * @date 1/31/2023 - 8:10:14 AM
 *
 * @param {useCalculatorProps} { value, setValue, operation, setOperation }
 * @returns {{ deleteValue: () => React.SetStateAction<string>=void; inputNumber: (number: React.ReactNode) => React.SetStateAction<string>=void;
 *  inputOperator: (operator: React.ReactNode) => void; }}
 */

const useCalculator = ({
  value,
  setValue,
  operation,
  setOperation,
}: useCalculatorProps) => {
  const operations = ["÷", "x", "-", "+", "%", "√"];

  /**
   * showToast()
   * Description: is a function that shows a native toast when the operator is incorrect
   * @date 1/29/2023 - 4:52:37 PM
   *
   * @param {string} message
   * @returns {ToastAndroid}
   */

  const showToast = (message: string) =>
    ToastAndroid.show(message, ToastAndroid.SHORT);

  /**
   * resetCalculator()
   * Description: This function reset the calculator values
   * @date 1/29/2023 - 4:53:38 PM
   *
   * @returns {void}
   */

  const resetCalculator = () => {
    setValue("0");
    setOperation("");
  };

  /**
   * deleteValue()
   * Description: This function delete the last value of the string value variable
   *  and reset the calculator when the value is 0
   * @date 1/29/2023 - 4:53:38 PM
   *
   * @returns {React.SetStateAction<string>=void}
   */

  const deleteValue = () => {
    const str = value.substring(0, value.length - 1);
    if (str === "0" || str.trim() === "") return resetCalculator();
    return setValue(str);
  };

  /**
   * inputNumber()
   * Description: This function set the calculator value when you make click on
   * the button
   * @date 1/29/2023 - 4:53:38 PM
   * @param {number} React.ReactNode
   * @returns {React.SetStateAction<string>=void}
   */

  const inputNumber = (number: React.ReactNode) => {
    const numberParsed = String(number);
    switch (numberParsed) {
      case "C":
        return resetCalculator();
      default: {
        if (value === "0") return setValue(numberParsed);
        else return setValue(`${value}${numberParsed}`);
      }
    }
  };

  /**
   * calculate()
   * Description: This function calculate the string operation with math.js library
   * but first replace the operators into another format supported by the library.
   * the button
   * @date 1/29/2023 - 4:53:38 PM
   *
   * @returns {void}
   */

  const calculate = () => {
    try {
      // Replace operators strings into another format supported by math.js (*, /, sqrt)
      const replacedValue = value
        .replace(/x/g, "*")
        .replace(/÷/g, "/")
        .replace(/√(\d+)/g, "sqrt($1)");
      // Evalute the operation
      const result = math.evaluate(replacedValue);
      // Parse into a string
      const resultParsed = String(result);
      // Set result
      setValue(resultParsed);
      setOperation("");
    } catch {
      showToast("Invalid format used.");
    }
  };

  /**
   * inputNumber()
   * Description: This function set the calculator value when you make click on
   * the button
   * @date 1/29/2023 - 4:53:38 PM
   *
   * @param {operator} React.ReactNode
   * @returns {void}
   */

  const inputOperator = (operator: React.ReactNode) => {
    const operatorParsed = String(operator);
    if (operations.includes(operatorParsed)) {
      let tmpValue = "";
      if (operation) {
        //Validate if the last element is a number to add another operator
        if (!isNaN(parseInt(value[value.length - 1]))) {
          tmpValue = `${value}${operatorParsed}`;
        } else {
          // Validate if is a square to concat with the operator
          if (operatorParsed === "√") {
            tmpValue = `${value}${operatorParsed}`;
          } else {
            // Change last operator by new operator
            let str = value.substring(0, value.length - 1);
            tmpValue = `${str}${operatorParsed}`;
          }
        }
      } else {
        // Validate if is a square to add instead 0
        if (value === "0" && operatorParsed === "√") {
          tmpValue = operatorParsed;
        } else {
          // Concat value calculator with operator
          tmpValue = `${value}${operatorParsed}`;
        }
      }
      // Set values
      setValue(tmpValue);
      setOperation(operatorParsed);
    }
    if (operatorParsed === "=") {
      // calculate
      calculate();
    }
  };

  return {
    deleteValue,
    inputNumber,
    inputOperator,
    resetCalculator,
    calculate,
  };
};

export default useCalculator;
