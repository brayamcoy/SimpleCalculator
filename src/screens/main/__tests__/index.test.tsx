import React from "react";
import useCalculator from "../../../hooks/useCalculator";
import Calculator from "../index";
import { create } from "react-test-renderer";

jest.mock("../../../hooks/useCalculator", () => {
  return jest.fn().mockImplementation(() => {
    return {
      deleteValue: jest.fn(),
      inputNumber: jest.fn(),
      inputOperator: jest.fn(),
    };
  });
});

jest.mock('@expo/vector-icons/build/vendor/react-native-vector-icons/lib/create-icon-set.js', () => {
  return () => '';
});

describe("Calculator component", () => {
  it("renders correctly", () => {
    const tree = create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("calls the inputNumber function when a number button is pressed", () => {
    const inputNumber = jest.fn();
    useCalculator.mockImplementation(() => {
      return {
        deleteValue: jest.fn(),
        inputNumber: inputNumber,
        inputOperator: jest.fn(),
      };
    });
    const tree = create(<Calculator />).root;
    tree.findByProps({ children: "9" }).props.onPress();
    expect(inputNumber).toHaveBeenCalled();
  });

  it("calls the inputOperator function when an operator button is pressed", () => {
    const inputOperator = jest.fn();
    useCalculator.mockImplementation(() => {
      return {
        deleteValue: jest.fn(),
        inputNumber: jest.fn(),
        inputOperator: inputOperator,
      };
    });
    const tree = create(<Calculator />).root;
    tree.findByProps({ children: "รท" }).props.onPress();
    expect(inputOperator).toHaveBeenCalled();
  });
});
