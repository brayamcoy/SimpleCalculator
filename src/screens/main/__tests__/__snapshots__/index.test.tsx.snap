// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component renders correctly 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  proxiedProperties={
    Object {
      "colors": Array [
        4280427042,
        4281545523,
      ],
      "endPoint": Array [
        1,
        0,
      ],
      "locations": undefined,
      "startPoint": Array [
        0,
        0,
      ],
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderRadius": 30,
        "elevation": 5,
        "flex": 1,
        "marginBottom": 20,
        "marginHorizontal": 10,
        "marginTop": 40,
        "overflow": "hidden",
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFF",
          "display": "flex",
          "elevation": 30,
          "flexDirection": "row-reverse",
          "height": 79,
          "overflow": "hidden",
          "paddingHorizontal": 28,
          "paddingVertical": 15,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
          "zIndex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#333",
            "fontSize": 29,
          }
        }
      >
        0
      </Text>
    </View>
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4282664004,
            4282664004,
          ],
          "endPoint": Array [
            1,
            0,
          ],
          "locations": undefined,
          "startPoint": Array [
            0,
            0,
          ],
        }
      }
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "justifyContent": "space-between",
            "margin": 10,
            "padding": 10,
            "width": "73%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row-reverse",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              <Text />
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              %
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              âˆš
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row-reverse",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              9
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              8
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              7
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row-reverse",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              6
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              5
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              4
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row-reverse",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              3
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              2
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              1
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row-reverse",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              C
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              .
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 5,
                  "padding": 35,
                },
                false,
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              0
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#F88934",
            "elevation": 5,
            "justifyContent": "space-between",
            "paddingVertical": 10,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "borderRadius": 5,
                "padding": 35,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontSize": 22,
              }
            }
          >
            Ã·
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "borderRadius": 5,
                "padding": 35,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontSize": 22,
              }
            }
          >
            x
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "borderRadius": 5,
                "padding": 35,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontSize": 22,
              }
            }
          >
            -
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "borderRadius": 5,
                "padding": 35,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontSize": 22,
              }
            }
          >
            +
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "borderRadius": 5,
                "padding": 35,
              },
              false,
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontSize": 22,
              }
            }
          >
            =
          </Text>
        </View>
      </View>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
</ViewManagerAdapter_ExpoLinearGradient>
`;
